cpp/src/lists/copying/gather.cu

```diff
@@ -0,0 +1,194 @@
+  }
+
+  __device__ result_type operator()(argument_type index)
+  {
+    // the "upper bound" of the span for a given offset is always offsets+1;
+    size_type const* upper_bound_start = offsets + 1;
+    // "step 1" from above
+    auto const bound =
+      thrust::upper_bound(thrust::seq, upper_bound_start, upper_bound_start + offset_count, index);
+    size_type offset_index = thrust::distance(upper_bound_start, bound);
+    // "step 2" from above
+    size_type offset_subindex = offset_index == 0 ? index : index - offsets[offset_index];
+    // "step 3" from above
+    return offset_subindex + base_offsets[offset_index];
+  }
+};
+
+/**
+ * @copydoc cudf::lists::detail::gather_list_leaf
+ *
+ */
+std::unique_ptr<column> gather_list_leaf(column_view const& column,
+                                         gather_data const& gd,
+                                         cudaStream_t stream,
+                                         rmm::mr::device_memory_resource* mr)
+{
+  // gather map iterator for this level (N)
+  auto gather_map_begin = thrust::make_transform_iterator(
+    thrust::make_counting_iterator<size_type>(0), list_gatherer{gd});
+  size_type gather_map_size = gd.gather_map_size;
+
+  // call the normal gather
+  auto leaf_column =
+    cudf::type_dispatcher(column.type(),
+                          cudf::detail::column_gatherer{},
+                          column,
+                          gather_map_begin,
+                          gather_map_begin + gather_map_size,
+                          // note : we don't need to bother checking for out-of-bounds here since
+                          // our inputs at this stage aren't coming from the user.
+                          false,
+                          stream,
+                          mr);
+
+  // the column_gatherer doesn't create the null mask because it expects
+  // that will be done in the gather_bitmask() step.  however, gather_bitmask()
+  // only happens at the root level, and by definition this column is a
+  // leaf.  so we have to generate the bitmask ourselves.
+  // TODO : it might make sense to expose a gather() function that takes a column_view and
+  // returns a column that does this work correctly.
+  rmm::device_buffer null_mask{0, stream, mr};
+  size_type null_count = column.null_count();
+  if (null_count > 0) {
+    auto list_cdv = column_device_view::create(column);
+    auto validity = cudf::detail::valid_if(
+      gather_map_begin,
+      gather_map_begin + gd.gather_map_size,
+      [cdv = *list_cdv] __device__(int index) { return cdv.is_valid(index) ? true : false; },
+      stream,
+      mr);
+
+    leaf_column->set_null_mask(std::move(validity.first), validity.second);
+  }
+
+  return std::move(leaf_column);
+}
+
+/**
+ * @copydoc cudf::lists::detail::gather_list_nested
+ *
+ */
+std::unique_ptr<column> gather_list_nested(cudf::lists_column_view const& list,
+                                           gather_data& gd,
```

Comment:
```suggestion
                                           gather_data const& gd,
```
This looks like it can be made const.

Reply:
Can't.  This is the buffer that ends up propagating the temporary buffer that we free up (in the call to make_gather_data).

Reply:
Ok. I did not notice the `std::move(gd.base_offsets)` when I looked at this before.